R/sheets_speedread.R

```diff
@@ -0,0 +1,221 @@
+#' https://docs.google.com/spreadsheets/d/SPREADSHEET_ID/export
+#'
+#' # examples of the query part
+#' ?format=csv
+#' ?format=csv?range=CELL_RANGE
+#' ?format=csv?gid=SHEET_ID
+#' ?format=csv?range=A1_range?gid=SHEET_ID
+#' ```
+#'
+#' The export URL supports the use of a bearer token in the Authorization
+#' header.
+#'
+#' Gist of various query parameters that work for the export URL, with special
+#' emphasis on the PDF format, which doesn't help us:
+#'   * <https://gist.github.com/Spencer-Easton/78f9867a691e549c9c70>
+#'
+#' @section The datasource URL:
+#'
+#' The "Chart Tools datasource protocol" is another way to access values in
+#' Sheets, designed to make Sheets play nicely with the Google Charts API. We
+#' can use the datasource URL as another way to get data as CSV (or JSON,
+#' etc.). It is more powerful than the export URL, in that it supports an
+#' entire SQL-like query language.
+#'
+#' The datasource URL does not appear to work with an Authorization header.
+#' Instead, we must send an access token as part of the query.
+#'
+#' Stack Overflow answer that links to various pages re: the "Chart Tools
+#' datasource protocol"
+#'   * <https://stackoverflow.com/a/33727897/2825349>
+#'
+#' The most important link is about "Query Source Ranges":
+#'   * <https://developers.google.com/chart/interactive/docs/spreadsheets#queryurlformat>
+#'
+#' Examples of datasource URLs
+#' ```
+#' # here is the base URL:
+#' https://docs.google.com/spreadsheets/d/SPREADSHEET_ID/gviz/tq
+#'
+#' # examples of the query part
+#' ?tqx=out:csv&sheet={sheet_name}
+#' ?range=A1:C4
+#' ?headers=N
+#' ?gid=N
+#' ````
+#'
+#' A readr issue that demonstrates sending an Authorization header
+#'   * <https://github.com/tidyverse/readr/issues/935>
+#'
+#' @inheritParams sheets_cells
+#' @param .url Whether to use the `"export"` (default) or `"datasource"` URL.
+#' @param .method Whether to process the CSV data with `vroom::vroom()`
+#'   (default) or `readr::read_csv()`.
+#' @param ... Passed along to `vroom::vroom()` or `readr::read_csv()`.
+#'
+#' @return A [tibble][tibble::tibble-package]
+#' @export
+#'
+#' @examples
+#' # Using the large NBA sheet from
+#' # https://github.com/tidyverse/googlesheets4/issues/122
+#' spreadsheet_id <- "1mnWcn7bd7obaXd05rnXrEtgzMBLdy7ctsYvlQM52W00"
+#' (ss <- as_sheets_id(spreadsheet_id))
+#'
+#' sheets_speedread(ss, .url = "export", .method = "vroom")
+#' sheets_speedread(ss, .url = "export", .method = "readr")
+#' sheets_speedread(ss, .url = "datasource", .method = "vroom")
+#' sheets_speedread(ss, .url = "datasource", .method = "readr")
+#'
+#' if (FALSE) {
+#' library(bench)
+#' bnch <- bench::mark(
+#'   sheets_speedread(ss, .url = "export", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "export", .method = "readr"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "readr"),
+#'   iterations = 5,
+#'   check = FALSE
+#' )
+#' bnch
```

Comment:
Shown above but will repeat here:

```
> bnch
# A tibble: 4 x 13
  expression                                                     min median
  <bch:expr>                                                   <bch> <bch:>
1 sheets_speedread(ss, .url = "export", .method = "vroom")     5.16s  5.47s
2 sheets_speedread(ss, .url = "export", .method = "readr")     4.18s  5.36s
3 sheets_speedread(ss, .url = "datasource", .method = "vroom") 4.48s  5.16s
4 sheets_speedread(ss, .url = "datasource", .method = "readr") 5.96s  6.34s
# … with 10 more variables: `itr/sec` <dbl>, mem_alloc <bch:byt>, `gc/sec` <dbl>,
#   n_itr <int>, n_gc <dbl>, total_time <bch:tm>, result <list>, memory <list>,
#   time <list>, gc <list>
```

Reply:
Why not always use vroom? In the future, readr will use vroom behind the scenes.

Reply:
FWIW this dataset is not nearly big enough to show a difference between reading methods. If you download it locally on my machine it takes .144 seconds to read with readr and 0.059 to read with vroom, but the times here are just dominated by the sheets API and downloading the data.

I don't think it matters much which method you use, google sheets has a maximum of 5 million rows, so most data is going to parse quickly. And any data that does approach the maximums will likely take too long to download to be practical.

Reply:
> Why not always use vroom? In the future, readr will use vroom behind the scenes.

Yeah @jimhester suggested that I favour vroom. For the initial experiment, I just wanted to compare all 4 combinations of CSV-parser and URL.